### PR TITLE
throttle: Minimal destructor fix for Luminous

### PR DESCRIPTION
**Pull request by adamemerson**
Get rid of the undefined behavior of destroying condition variables while they're being waited on

### CODE REVIEW

File: src/common/Throttle.cc [hunk 1]
```diff
@@ -629,7 +654,9 @@ int OrderedThrottle::wait_for_ret() {
   complete_pending_ops();
 
   while (m_current > 0) {
+    ++waiters;
     m_cond.Wait(m_lock);
+    ++waiters;
```
Comment (gregsfortytwo) [inline src/common/Throttle.cc]: --

File: src/common/Throttle.cc [hunk 2]
```diff
@@ -95,26 +103,31 @@ bool Throttle::_wait(int64_t c)
...
+      ldout(cct, 2) << "_wait waiting..." << dendl;
+      if (logger)
+	start = ceph_clock_now();
+
+      do {
+	(*cv)->Wait(lock);
+      } while ((_should_wait(c) || cv != cond.begin()) && !shutting_down);
+      if (shutting_down) {
+	shutdown_clear.SignalOne();
+	throw waiting_destructed("Throttle destructed while we waited.");
```
Comment (gregsfortytwo) [inline src/common/Throttle.cc]: Uh, we don't use exceptions.[1] Is there some reason not to just make this an assert?

[1]: (Except that we do in bufferlist decode, but that's the only place unless something's changed, and they're outlawed by https://google.github.io/styleguide/cppguide.html#Exceptions)

File: src/common/Throttle.cc (hunk 1 above)
Comment (adamemerson) [inline src/common/Throttle.cc]: WELL, that's what I get for copying my diff without paying close enough attention.

**Comment by adamemerson** [comment]
I just missed SimpleThrottle when I was moving my patch over, I'll catch that in a moment.

I think Throttle is the only one that has a test for this case. I can write ones for the others.

**Comment by adamemerson** [comment]
(Death Tests are across threads are pretty hairy, though.)

**Comment by adamemerson** [comment]
Ah, when I was doing this last night I thought the m_current assertion in SimpleThrottle checked for waiters, but I was mistaken about that. One moment.

**Comment by adamemerson** [comment]
I didn't know about scope_guard either until Kefu pointed out that the 'ward' function I wrote duplicated its functionality :)

File: src/common/Throttle.cc (hunk 2 above)
Comment (gregsfortytwo) [inline src/common/Throttle.cc]: Mostly we don't like the way it messes around with control flow. We also use the google style guide as our baseline and nobody has ever made a persuasive argument for breaking with it in this case.

(Also, it's polite to users of our libraries not to make them play with exceptions.)

**Comment by adamemerson** [comment]
There, I think I've addressed everything.

**Comment by gregsfortytwo** [comment]
These make check errors are getting tedious, but they look like our current noise. :(

Reviewed-by: Greg Farnum <gfarnum@redhat.com>

**Comment by adamemerson** [comment]
What the heck, Jenkins? Retest this, please.

**Comment by adamemerson** [comment]
Why are you doing this to me, Jenkins? Retest this, if you please.

**Comment by adamemerson** [comment]
Jenkins, retest this please.

File: src/test/common/Throttle.cc [hunk 3]
```diff
@@ -58,6 +59,8 @@ class ThrottleTest : public ::testing::Test {
     }
   };
 
+  static void destruct_throttle() {
```
Comment (tchaikov) [inline src/test/common/Throttle.cc]: @adamemerson what is this method for? did you plan to extract the common piece into it and reuse it?

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-07-29_05:31:06-rados-wip-kefu-testing-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-07-30_00:01:51-rados-wip-kefu-testing-distro-basic-smithi/

**Comment by adamemerson** [comment]
Does this mean I can merge? I'm never quite sure how many of the random 'dead' and 'failed' things on Pulpito are blockers and how many are 'Our QA system is flaky'.

**Comment by adamemerson** [comment]
@tchaikov It's a leftover. I'll get rid of it. I had originally meant to templatize the test function and reuse it, but all the throttles have different interfaces so that didn't work.

**Comment by tchaikov** [comment]
@adamemerson yeah, i think it will be ready to merge once `destruct_throttle()` is removed. i checked all the test failures. they are either known issues or issues being addressed by PRs not yet merged. i was just too lazy to list them all.

- http://tracker.ceph.com/issues/20700: ceph-objectstore-tool: exp list-pgs failure with status 1
- https://github.com/ceph/ceph/pull/16680, and https://github.com/ceph/ceph/pull/16706: Command failed ... crush_ops.sh
- https://github.com/ceph/ceph/pull/16663: Command failed  ... test_failure.sh, that PR failed this test, so it's not relevant.
- http://tracker.ceph.com/issues/20854: a new issue. again, i don't think it's relevant.

**Pull request closed without merging**